tests/integration/standard/test_shard_aware.py

```diff
@@ -0,0 +1,106 @@
+# Copyright DataStax, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+try:
+    import unittest2 as unittest
+except ImportError:
+    import unittest  # noqa
+
+from cassandra.cluster import Cluster
+from cassandra.policies import TokenAwarePolicy,RoundRobinPolicy
+from cassandra.query import SimpleStatement
+
+class TestShardAwareIntegration(unittest.TestCase):
+    def test_basic(self):
+        cluster = Cluster(['127.0.0.1'], protocol_version=4, load_balancing_policy=TokenAwarePolicy(RoundRobinPolicy()))
+        session = cluster.connect()
+
+        session.execute(
+            """
+            DROP KEYSPACE IF EXISTS preparedtests
+            """
+        )
+        session.execute(
+            """
+            CREATE KEYSPACE preparedtests
+            WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}
+            """)
+
+        session.set_keyspace("preparedtests")
+        session.execute(
+            """
+            CREATE TABLE cf0 (
+                a text,
+                b text,
+                c text,
+                PRIMARY KEY (a, b)
+            )
+            """)
+
+        prepared = session.prepare(
+            """
+            INSERT INTO cf0 (a, b, c) VALUES  (?, ?, ?)
+            """)
+
+        bound = prepared.bind(('a', 'b', 'c'))
+
+        session.execute(bound)
+
+        bound = prepared.bind(('e', 'f', 'g'))
+
+        session.execute(bound)
+
+
+        bound = prepared.bind(('100000', 'f', 'g'))
+
+        session.execute(bound)
+
+        prepared = session.prepare(
+            """
+            SELECT * FROM cf0 WHERE a=?
+            """)
+
+        bound = prepared.bind(('a'))
+        results = session.execute(bound, trace = True)
+
+        events = results.get_query_trace().events
+        for event in events:
+            print(event.thread_name, event.description)
+        
+        for event in events:
+            self.assertEqual(event.thread_name, "shard 4")
```

Comment:
Please, provide some explanation why this works... ;)

Reply:
I did had a question regarding this, cause running this multiple time, I get some case that the query goes also to a different shard is that normal ? or there is anything else to explain that ?

Reply:
If you could explain the logic of your test it would be easier for me to relate.
Also provide the tracing results that you think make no sense together with those that DO make sense.

Reply:
this is the query (the last one in the test)
```python
        prepared = session.prepare(
            """
            SELECT * FROM cf0 WHERE a=?
            """)

        bound = prepared.bind(('e'))
        results = session.execute(bound, trace = True)
```
this is the working example, when all the traces are from shard 4: 
```
# query 
shard 4 Checking bounds
shard 4 Processing a statement
shard 4 Creating read executor for token -4200008757497435756 with all: {172.17.0.2} targets: {172.17.0.2} repair decision: NONE
shard 4 read_data: querying locally
shard 4 Start querying the token range that starts with -4200008757497435756
shard 4 Querying is done
shard 4 Done processing - preparing a result
```

the test failing, when shard 8 is approached 
```
shard 4 Processing a statement
shard 4 Creating read executor for token -4200008757497435756 with all: {172.17.0.2} targets: {172.17.0.2} repair decision: NONE
shard 4 read_data: querying locally
shard 4 Start querying the token range that starts with -4200008757497435756
shard 4 Querying is done
shard 4 Done processing - preparing a result
shard 4 Parsing a statement
shard 4 Processing a statement
shard 4 Creating read executor for token -4869684399737000146 with all: {172.17.0.2} targets: {172.17.0.2} repair decision: NONE
shard 4 read_data: querying locally
shard 8 Start querying the token range that starts with -4869684399737000146
shard 8 Querying is done
shard 4 Done processing - preparing a result
```

I'm not sure we there another read with token -4869684399737000146...

Reply:
The last one contains traces from at least two CQL statements.
How do you get tracing info?
Let's have a call in order to speed up the convergence of this...

Reply:
I think I've found the issue, seem that query is actually reading the tracing, which seem so be a bug in tracing,
after reading only the trace_id, and query it separately, the test pass consistently....

Reply:
What kind of bug in tracing you have in mind? On the driver side? 
That would be weird - driver is supposed to do exactly what you described above: read events that correspond to a particular `session_id`.

Reply:
I've raised it on https://github.com/scylladb/scylla/issues/6004

I think it's a scylla bug, that in some case it append this query into the trace events table

Reply:
I've now even think I've proven to myself it a scylla bug that I see in 3.2.0, and not on master.